protos/idl/service.proto

```diff
@@ -98,3 +116,74 @@ message Tag {
 message Metadata {
     map<string, string> key_map = 1; // key map is a dictionary of key/val strings that represent metadata
 }
+
+// Filter expression that is composed of a combination of single filters
+message FilterExpression {
+    repeated SinglePropertyFilter filters = 1;
+}
+
+// A single property to filter on.
+message SinglePropertyFilter {
+    oneof property_filter {
+        TagPropertyFilter tag_filter = 1;
+        PartitionPropertyFilter partition_filter = 2;
+        ArtifactPropertyFilter artifact_filter = 3;
+        DatasetPropertyFilter dataset_filter = 4;
+    }
+
+    // as use-cases come up we can add more operators, ex: gte, like, not eq etc.
+    enum ComparisonOperator {
+        EQUALS = 0;
+    }
+
+    ComparisonOperator operator = 10; // field 10 in case we add more entities to query
+    // Next field number: 11
+}
+
+// Artifact properties we can filter by
+message ArtifactPropertyFilter {
+    oneof property {
+        string artifact_id = 1;
+    }
+}
+
+// Tag properties we can filter by
+message TagPropertyFilter {
+    oneof property {
+        string tag_name = 1;
+    }
+}
+
+// Partition properties we can filter by
+message PartitionPropertyFilter {
+    oneof property {
+        KeyValuePair key_val = 1;
+    }
+}
+
+message KeyValuePair {
+    string key = 1;
+    string value = 2;
+}
+
+// Dataset properties we can filter by
+message DatasetPropertyFilter {
+    oneof property {
+        string project = 1;
+        string name = 2;
+        string domain = 3;
+        string version = 4;
+    }
+}
+
+message PaginationOptions {
+    int32 limit = 1;
+    int32 token = 2;
+    string sortKey = 3;
```

Comment:
what's a sortKey used for?

Reply:
`sortKey` is the values property to sort by. For example `CreatedAt` or `UpdatedAt`.

Reply:
might be nice to add a comment unless i'm just being particularly dense today (blaming pagerduty for the sleep deprivation :P)

Reply:
oh jk you have!